src/amr_criteria/refinement_package.cpp

```diff
@@ -150,9 +157,117 @@ AmrTag SecondDerivative(const AMRBounds &bnds, const ParArray3D<Real> &q,
   return AmrTag::same;
 }
 
-void SetRefinement_(MeshBlockData<Real> *rc) {
+void FirstDerivative(const AMRBounds &bnds, MeshData<Real> *md, const std::string &field,
+                     const int &idx, ParArray1D<AmrTag> &amr_tags,
+                     const Real refine_criteria_, const Real derefine_criteria_,
+                     const int max_level_) {
+  const auto desc =
+      MakePackDescriptor(md->GetMeshPointer()->resolved_packages.get(), {field});
+  auto pack = desc.GetPack(md);
+  const int ndim = md->GetMeshPointer()->ndim;
+  const int nvars = pack.GetMaxNumberOfVars();
+
+  const Real refine_criteria = refine_criteria_;
+  const Real derefine_criteria = derefine_criteria_;
+  const int max_level = max_level_;
+  const int var = idx;
+  // get a scatterview for the tags that will use Kokkos::Max as the reduction operation
+  auto scatter_tags = amr_tags.ToScatterView<Kokkos::Experimental::ScatterMax>();
+  par_for_outer(
+      PARTHENON_AUTO_LABEL, 0, 0, 0, pack.GetNBlocks() - 1, bnds.ks, bnds.ke, bnds.js,
+      bnds.je,
+      KOKKOS_LAMBDA(team_mbr_t team_member, const int b, const int k, const int j) {
+        Real maxd = 0.;
+        par_reduce_inner(
+            inner_loop_pattern_ttr_tag, team_member, bnds.is, bnds.ie,
+            [&](const int i, Real &maxder) {
+              Real scale = std::abs(pack(b, var, k, j, i));
+              Real d = 0.5 *
+                       std::abs((pack(b, var, k, j, i + 1) - pack(b, var, k, j, i - 1))) /
+                       (scale + TINY_NUMBER);
+              maxder = (d > maxder ? d : maxder);
+              if (ndim > 1) {
+                d = 0.5 *
+                    std::abs((pack(b, var, k, j + 1, i) - pack(b, var, k, j - 1, i))) /
+                    (scale + TINY_NUMBER);
+                maxder = (d > maxder ? d : maxder);
+              }
+              if (ndim > 2) {
+                d = 0.5 *
+                    std::abs((pack(b, var, k + 1, j, i) - pack(b, var, k - 1, j, i))) /
+                    (scale + TINY_NUMBER);
+                maxder = (d > maxder ? d : maxder);
+              }
+            },
+            Kokkos::Max<Real>(maxd));
+        auto tags_access = scatter_tags.access();
+        auto flag = AmrTag::same;
+        if (maxd > refine_criteria && pack.GetLevel(b, 0, 0, 0) < max_level)
```

Comment:
Why has the `max_level` condition been introduced here (and for the other second deriv)?
Is this actually required (because it's also checked in `SetRefinement` IIRC.
And if it's needed, does this mean downstream code would also have to include this kind of check in custom refinement ops?

Reply:
I see. That makes sense.

I'm still not sure this check should belong inside the kernel (rather than being a general feature of those refinement criteria derived from what's available within the refinement package).
Conceptually, I see the refinement critera itself as what's inside the kernel and the max level an outer constraint (though also being available on a per refinement criteria level).
So I see a danger if there are more refinement criteria are added the code for checking the max level would need to be duplicated and added to each kernel, rather than checking separately.

Having said that, I'm happy to fix this later so that this PR can go in quicker.